### ISSUE TITLE
HTTP grammar

### ISSUE THREAD

**Issue by cppden** [taocpp/PEGTL#174]
HTTP grammar seems not viable for chunked body. 
RFC describes the chunk as:
```
chunk      = chunk-size [ chunk-ext ] CRLF chunk-data CRLF
chunk-data = 1*OCTET ; a sequence of chunk-size octets
```
The tricky part is in comments. Basically it says we should skip chunk-size octets w/o actually parsing them. PEGTL generously provided implementation of HTTP grammar like this:
```cpp
struct chunk_data : until< at<CRLF>, OCTET > {};
struct chunk : seq< chunk_size, opt< chunk_ext >, CRLF, chunk_data, CRLF > {};
```
But we can't parse chunk-body till CRLF since the data inside chunk can be anything including CRLF. We need to blindly skip that many octets and then resume parsing.

Is there any way in PEGTL to properly do this? In theory we should pass the result of chunk_size into parser FSM to instruct it on skipping the requested number of bytes...

One ore time, thank you guys for such handy and powerful library!

**Comment by ColinH** [comment]
We can solve that with a custom rule or two, though it will require some fiddling around with the states. I'll look into it as soon as I have time.

**Comment by ColinH** [comment]
The above commit changes the `chunk_size` and `chunk_data` rules in a temporary copy of `http.hpp` called `http_chunk_size.hpp` to work as expected, however they require a `size_t` as first State argument to pass the size from one rule to the other.

The next step is to change the `chunk` rule in a way that the extra `size_t` is automatically added to the States where required, and hidden where its not (instead of the temporary, highly intrusive solution from yesterday where the user needs to take care of it).

**Comment by ColinH** [comment]
The state handling is now implemented, and tested, which also helped me find one or two other minor issues in the grammar. If nothing else comes up we'll replace `http.hpp` with the temporary `http_chunk_size.hpp` which should conclude this operation.

@d-frey If you could look at the current implementation and check whether I missed something obvious. Then we should also check whether something generic can be extracted and placed in a dedicated header file to make other stateful rules easier to implement.

**Comment by ColinH** [comment]
@cppden Can you confirm that this works for you, too?

**Comment by cppden** [comment]
@ColinH, thanks a lot for the fix! I can't yet follow its machinery... Perhaps it's possible to somehow provide a generic means in your lib for such things (only if possible to allow mortals to use)?
A simple UT with CRLF in chunk data would do the check (e.g. from [wiki](https://en.wikipedia.org/wiki/Chunked_transfer_encoding). I'm sorry but can't use the current grammar in production because of #175. For now I'm using pieces of the whole grammar with external FSM tracking the input and calling parser only when enough data received.

**Comment by ColinH** [comment]
I'm assuming that the hard to follow part of the machinery is the state-hiding, i.e. where we make sure that the additional `std::size_t` State that the `chunk_size` rule needs to tell the `chunk_data` the size of the chunk is hidden from all user actions etc.

As you can see we have factored out some of the puzzle pieces, and while it's sufficient for this particular use case we'll also see what other related things could be made generic and offered pre-built.

Regarding #175 it's the same here, in a real-world application there's no real alternative than to parse the chunk-header with the PEGTL and then just read that many bytes from the socket. Until we have standard co-routines...
